### ISSUE TITLE
Arrow keys do not change value of select

### ISSUE THREAD

**Issue by CUnknown** [select2/select2#3472]
According to https://github.com/select2/select2/issues/3279 the feature of opening the select2 search dropdown is not implemented in the newer version according to technical difficulties, but also because it does not reflect the default behaviour of a native select.

Though, following this native select behaviour, the user should be able to browse through the dropdown items by using the arrow keys on the keyboard. This does not seem to work for the select2 4.0 implementation. Is there a reason for this?

**Comment by AlexBuitrago** [comment]
+1000

**Comment by rain01** [comment]
yes please

**Comment by sidharthc** [comment]
Is there a solution for this?

**Comment by uzlov** [comment]
this.on('keypress', function (evt) {
      var key = evt.which;

      if (self.isOpen()) {
        if (key === KEYS.ENTER) {
          self.trigger('results:select');

          evt.preventDefault();
        } else if ((key === KEYS.SPACE && evt.ctrlKey)) {
          self.trigger('results:toggle');

          evt.preventDefault();
        } else if (key === KEYS.UP) {
          self.trigger('results:previous');

          evt.preventDefault();
        } else if (key === KEYS.DOWN) {
          self.trigger('results:next');

          evt.preventDefault();
        } else if (key === KEYS.ESC || key === KEYS.TAB) {
          self.close();

          evt.preventDefault();
        }
      } else {
        if (key === KEYS.ENTER || key === KEYS.SPACE ||
            ((key === KEYS.DOWN || key === KEYS.UP) && evt.altKey)) {
          self.open();

          evt.preventDefault();
        }
        if (key === KEYS.DOWN) {
          this.$element.find('option:selected').next().attr('selected', 'selected').trigger('change');
          evt.preventDefault();
        }
        if (key === KEYS.UP) {
          this.$element.find('option:selected').prev().attr('selected', 'selected').trigger('change');
          evt.preventDefault();
        }
      }
    });

**Comment by rain01** [comment]
@uzlov what is this? where do you put it?

**Comment by sidharthc** [comment]
@uzlov seems good.

**Comment by uzlov** [comment]
@rain01
it's a patch for select2
please, find function this.on('keypress', function (evt) {
and replace with my code

**Comment by alvarotrigo** [comment]
+1

**Comment by CUnknown** [comment]
@uzlov could you submit a PR?

**Comment by alvarotrigo** [comment]
I believe the change have to be made [in line `323` of core.js](https://github.com/select2/select2/blob/master/src/js/select2/core.js#L323).

An easy change that improves the usability and accessibility of the plugin. In my case my boss wouldn't allow me to use it unless this feature was provided. 

[Still need to fix the autofocus](https://github.com/select2/select2/issues/3696), but hopefully I'll manage to do it.

**Comment by alvarotrigo** [comment]
Or if you prefer, [line 5150](https://github.com/select2/select2/blob/master/dist/js/select2.js#L5150) of the full distributed file.

**Comment by sauron918** [comment]
+1

**Comment by Kortsmit** [comment]
+1

**Comment by bradparks** [comment]
For sure, +1. I'm surprised this is an issue still !

**Comment by Decagon** [comment]
+1

**Comment by wsromek** [comment]
:+1:

**Comment by WimDeMeester** [comment]
@uzlov 's code is working fine for me. Any change that this will be included in the next version (4.0.2?) of select2?

**Comment by phazei** [comment]
I've been using @uzlov code, rather than modifying the code I just go:
```javascript
                    s2Obj = $(element).data('select2');
                    delete s2Obj.listeners.keypress;
                    s2Obj.on('keypress', function (evt) {
                        //his code here
                    });
```

But I also just found an issue with it.  If the item in the select list is disabled, and you arrow to it, it actually selects it instead of going to the next valid option.  It doesn't really select it, but it shows in the box.  My validation marks it as invalid so it's sort of ok, but it should be next active item.

**Comment by phazei** [comment]
Note: in @uzlov code, the KEYS.DOWN and KEYS.UP items should be changed to only select the enabled options:

```javascript
                            if (key === KEYS.DOWN) {
                                var val = s2Obj.$element.find('option:selected').nextAll(":enabled").first().val();
                                if (undefined !== val) {
                                    s2Obj.$element.val(val);
                                    s2Obj.$element.trigger('change');
                                }
                                evt.preventDefault();
                            }
                            if (key === KEYS.UP) {
                                var val = s2Obj.$element.find('option:selected').prevAll(":enabled").first().val();
                                if (undefined !== val) {
                                    s2Obj.$element.val(val);
                                    s2Obj.$element.trigger('change');
                                }
                                evt.preventDefault();
                            }
```

**Comment by lucjan** [comment]
+1

**Comment by alexweissman** [comment]
I believe either #4293 or #5130 will resolve this issue.

**Comment by aaime** [comment]
+1

**Comment by KyleMit** [comment]
It's probably best to not modify the [library contents][0] ad-hoc (and especially difficult if you're serving from a CDN).  You can implement [uzlov's suggestion][1] (and [phazei's update][2]) by attaching a listener to the object after the library has already loaded.  You can listen for key strokes on the `.select2` input when it's closed or the `.select2-dropdown` if it's open.  And then traverse the dom to find the element and update accordingly after the fact like this:

```js
 // listen for keyups in both input widget AND dropdown
$("body").on('keyup', ".select2,.select2-dropdown", function (e) {
  var KEYS = { UP: 38, DOWN: 40 };
  var $sel2 = $(this).closest(".select2");
  // if we can't find it by traveersing the dom, search page for an open container
  // ASSUMES only one open at a time - 
  // don't really know how to cross walk from dropdown back to coresponding element
  if ($sel2.length == 0) {
  	$sel2 = $(".select2.select2-container--open");
  }

  // make sure we found the <select> el
  var $sel = $sel2.data("element")
  if ($sel.length) {
    var newValue

    if (e.keyCode === KEYS.DOWN && !e.altKey) {
      newValue = $sel.find('option:selected').nextAll(":enabled").first().val();
    } else if (e.keyCode === KEYS.UP) {
      newValue = $sel.find('option:selected').prevAll(":enabled").first().val();
    }

    // if we got a value, set it and update
    if (newValue != undefined) {
      $sel.val(newValue);
      $sel.trigger('change');
    }
  }

});
```

### [Demo in jsFiddle][3]

[0]: https://github.com/select2/select2/blob/4.0.5/src/js/select2/core.js#L319
[1]: https://github.com/select2/select2/issues/3472#issuecomment-132246474
[2]: https://github.com/select2/select2/issues/3472#issuecomment-187941393
[3]: https://jsfiddle.net/KyleMit/nb43Lma7/

**Comment by kfrz** [comment]
@KyleMit this is a nice solution, the only thing I've noticed is there's no `e.preventDefault()` which means when selecting elements with arrow key, the page will scroll -- however I cannot seem to figure where to place the `preventDefault()` to make it work. I'm also of the "I don't want to to modify core lib code" camp, so until https://github.com/select2/select2/pull/4293 is merged this workaround will suffice.

**Comment by devonwaldon-pl** [comment]
@kfrz What worked for me was changing over to a `keydown` event, instead of the `keyup` suggested above.

**Comment by avernet** [comment]
+1

**Comment by B8ightY** [comment]
This is still an issue
+1

**Comment by Jagdish-J-P** [comment]
Any update on this?

**Comment by zpiecuch81** [comment]
Any update on this?
I've just hit similar issue. Was it fixed in any version?

**Comment by PBrockmann** [comment]
Any news on this issue ? Have not been able to find a solution with 4.0.13 release.
Crudely missing this default feature that is available with a classic select !

**Comment by PBrockmann** [comment]
Here is a siple solution that runs with select2 4.0.13

```
$("body").on('keyup', ".select2", function(e) {
  var KEYS = {UP: 38, DOWN: 40};
  var sel = $(this).closest('.select2-container').parent().find('select');
  if (e.keyCode === KEYS.DOWN) {
    newValue = $(sel).find(':selected').next().val();
  } else if (e.keyCode === KEYS.UP) {
    newValue = $(sel).find(':selected').prev().val();
  }
  if (newValue != undefined) {
    $(sel).val(newValue).trigger('change');
  }
});
```

The jsfiddle is available from https://jsfiddle.net/PBrockmann/ktcnr6o7/

**Comment by nickcoast** [comment]
@PBrockmann That works great. The only issues I had were:
1. The entire viewport scrolls when hitting UP and DOWN arrows. Adding `e.preventDefault()` fixes that for me.
2. Pressing and holding UP or DOWN only scroll by one value, instead of scrolling continuously. So I changed 'keyup' to 'keydown'.

```

$("body").on('keydown', ".select2", function(e) { // changed to 'keydown'
  var KEYS = {UP: 38, DOWN: 40};
  var sel = $(this).closest('.select2-container').parent().find('select');
  var newValue;

  if (e.keyCode === KEYS.DOWN) {
    e.preventDefault(); // Prevent the whole viewport from scrolling
    newValue = $(sel).find(':selected').next().val();
  } else if (e.keyCode === KEYS.UP) {
    e.preventDefault(); // Prevent the whole viewport from scrolling
    newValue = $(sel).find(':selected').prev().val();
  }
  if (newValue !== undefined) {
    $(sel).val(newValue).trigger('change');
  }
});
```

**Comment by PBrockmann** [comment]
Thank you for the feedback but I do not have the same behaviour with google-chrome. 
No need to add `e.preventDefault()`.

Could you precise what browser is concerned ?

**Comment by gorillapower** [comment]
Thank you for creating this solution. 

In my case, I needed to
1. Trigger the `select2:select` event and pass in original data object with custom properties
2. Check for disabled options

See the additions below.

``` javascript

$("body").on('keydown', ".select2", function(e) { // changed to 'keydown'
    var KEYS = {
        UP: 38, 
        DOWN: 40
    };
    
    var sel = $(this).closest('.select2-container').parent().find('select');
    var newValue;
    
    if (e.keyCode === KEYS.DOWN) {
        e.preventDefault(); // Prevent the whole viewport from scrolling
        newValue = $(sel).find(':selected').next(":not([disabled])").val(); //NEW: gets next not disabled option
    } else if (e.keyCode === KEYS.UP) {
        e.preventDefault(); // Prevent the whole viewport from scrolling
        newValue = $(sel).find(':selected').prev(":not([disabled])").val(); //NEW: gets next not disabled option
    }
  
    
    if (newValue !== undefined) {
        $(sel).val(newValue).trigger('change');
        
        //NEW: Add data object with custom properties
        var data = $(sel).select2('data')[0];
        $(sel).trigger({
            type: 'select2:select',
            params: {
                data: data
            }
        });
    }
  });

```

**Comment by shapeh** [comment]
Was wondering if someone could make a PR for this, so we can close the matter. 
This issue was reported in 2015. Shall we celebrate it's 10th year birthday next year?
